### PR TITLE
update

### PR DESCRIPTION
**Pull request by Rich-Harris**
Alternative to #134. This is more or less a rewrite — it uses all the new idioms we've been adding over the last few months, and it's been an absolute joy

### CODE REVIEW

**Review by benmccann — COMMENTED** [review]
https://github.com/sveltejs/realworld/pull/134 is quite obviously a steaming pile of crap in many ways and not something I wanted to merge, so I'm glad to see some cleanup for this app. There are a couple things I'd like to be able to do that drove a lot of that ugliness. They're not necessarily all that important for this app, but do shine light on limitations we have that have been affecting me outside this app as well. I'd really like to see us come up with better solutions for a couple of these issues and I was using #134 as a bit of a test ground for that.

One is that it's really hard to keep information in stores to be able to update layout elements without refetching the whole page. In this app, for example, it would be nice if we could have the `user` info in a store so that if you update the user info on the profile page the `username` in `Nav.svelte` can be updated without having to refresh all the data on the page. Fetching all the data on the page again isn't the worst thing in the world for this app, but you can imagine cases where that's more painful and it seems like something you should be able to avoid without it being as hard as it is.

Another thing was around login/logout. E.g. if you logout with this PR right now it first fetches `/settings/_data.json` and then fetches `/login/_data.json`. It seems unnecessary to fetch both. We should just be able to delete the cookie and then navigate to the new page without having to first refresh the current page

**Comment by yadoga** [comment]
Realworld has been a valuable reference (JWT auth etc.) and I frequently refer others to it. Commenting out of appreciation 🙌

File: src/routes/settings/+page.server.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+export async function load({ locals }) {
+	if (!locals.user) throw redirect(302, '/login');
+}
+
+/** @type {import('./$types').Actions} */
+export const actions = {
+	logout: async ({ cookies, locals }) => {
+		cookies.delete('jwt', { path: '/' });
+		locals.user = null;
```
[the remark this answers is not in the document]
Comment (Rich-Harris) [reply src/routes/settings/+page.server.js:12]: It does. See https://github.com/sveltejs/kit/issues/7482

File: src/routes/settings/+page.svelte [hunk 2]
```diff
@@ -39,68 +19,68 @@
...
 						</fieldset>
 
-						<button
-							class="btn btn-lg btn-primary pull-xs-right"
-							type="submit"
-							disabled={in_progress}
-						>
-							Update Settings
-						</button>
+						<button class="btn btn-lg btn-primary pull-xs-right"> Update Settings </button>
```
[the remark this answers is not in the document]
Comment (Rich-Harris) [reply src/routes/settings/+page.svelte:75]: no, that's just what Prettier does if you have the text on a separate line to the tags

File: svelte.config.js [hunk 3]
```diff
@@ -1,10 +1,7 @@
-import adapter from '@sveltejs/adapter-auto';
+import adapter from '@sveltejs/adapter-vercel';
```
[the remark this answers is not in the document]
Comment (Rich-Harris) [reply svelte.config.js:1]: it was failing to build on netlify and i couldn't be bothered to figure out why (some caching issue i think) so i figured it'd be easier to move it somewhere... better. deleted the config file

**Comment by Rich-Harris** [comment]
> One is that it's really hard to keep information in stores to be able to update layout elements without refetching the whole page.

In the vast majority of cases I think you're better off having a single source of truth, i.e. the server. But if you do need to do that, you can just return a store from a `load` function, no?

**Comment by Rich-Harris** [comment]
> Another thing was around login/logout. E.g. if you logout with this PR right now it first fetches `/settings/_data.json` and then fetches `/login/_data.json`. It seems unnecessary to fetch both. We should just be able to delete the cookie and then navigate to the new page without having to first refresh the current page

The details of what happens depends on whether it's a native or enhanced submission:

* native — we run all the `load` functions again, including the one in `src/routes/settings/+page.server.js`, at which point the redirect is thrown and SvelteKit returns a 302 response (should probably be a 307 but whatever). The browser reloads the page to `/login`
* enhanced — we run all the `load` functions again, hit the redirect, and `/settings/__data.json` contains instructions to redirect. SvelteKit navigates client-side to `/login`, requesting data from the server as it does so

Admittedly there's some room for micro-optimisation here — if we redirected at the end of the `logout` action we wouldn't need to run the `load` functions again. In practice that wouldn't make any difference, since both applicable `load` functions are synchronous and don't do any real work. 

There's obviously nothing we can do to improve the native case.

In the enhanced case, we _could_ recognise that a redirect had been thrown, and load data for the target page eagerly. In other words instead of `/settings/__data.json` looking like this...

```js
{
  "type": "redirect",
  "location": "/login"
}
```

...it could look like this:

```js
{
  "type": "redirect",
  "location": "/login",
  "nodes": [
    { "type": "data", "data": { "user": null } },
    null
  ]
}
```

That would be a neat optimisation, but there are two caveats:

1. it would add some bytes complexity to the SvelteKit client, though perhaps not an unbearable amount
2. there's no guarantee that the user wants to apply the redirect — they could be returning a custom function in their `enhance` callback. If they didn't redirect, we'd have loaded that data (and increased the payload) unnecessarily. So we'd probably want to only do this when `use:enhance` is being used but no custom function was used. Still a win, but more limited in scope

**Review by dummdidumm — APPROVED** [review]
I only skimmed the code and played around with the app a little, from that it looks good. I'm wondering if we can do something to get the amount of CSS down a little, right now we're loading the whole of IonicIcons and BootStrap, which worsens the lighthouse score a bit. Definitely not something for this PR, but something to note either way.

**Comment by Rich-Harris** [comment]
Yeah, the CSS could definitely be optimised. We're basically following the instructions from https://realworld-docs.netlify.app/docs/specs/frontend-specs/templates — not sure if it would be 'cheating' to optimise it

**Pull request merged**